[Discover] Check if user has any access to any registered resource

kimlisa opened the pull request:
#### Description
part of https://github.com/gravitational/teleport/issues/13997

- adds an endpoint for web apiserver to check if user has any access to a registered resource
- added the new connection diagnostic access perms to UI user context
- tweaked list resources to not always require fake pagination

kimlisa committed:
Add connection diagnostic access to UI user context

kimlisa committed:
Implement resource checker and tweak list resources

marcoandredinis commented on lib/web/ui/usercontext.go:
```diff
@@ -83,6 +83,8 @@ type userACL struct {
 	// Billing defines access to billing information.
 	Billing access `json:"billing"`
+	// ConnectionDiagnostic defines access to connection diagnostics.
+	ConnectionDiagnostic access `json:"connectionDiagnostic"`
```
Is this change related with the check we are doing?

kimlisa replied:
i didn't know how literally to take this question :sweat_smile: so i provided both

yes, it's related in that it's required for discovery, so we can pre-check that users have all the necessary perms before allowing user to start the discovery flow

no, in that it's not required for the `does user has at least 1 resource` check, i just sneaked it in b/c it was tiny add in (maybe i should've have?)

marcoandredinis replied:
Got it :+1: 
Thank you

marcoandredinis commented on lib/web/resources_test.go:
```diff
@@ -375,7 +375,6 @@ func TestListResources(t *testing.T) {
 			expected: proto.ListResourcesRequest{
 				ResourceType:        types.KindNode,
 				Limit:               defaults.MaxIterationLimit,
-				NeedTotalCount:      true,
```
Why are we removing this field?

kimlisa replied:
so before, i set this field to always be true because i thought the list resources request coming through the web api layer will always require it (bad decision on my part), but this field isn't actually necessary because our frontend calls with the query `sort` always set, which will also return the total count.

marcoandredinis commented on lib/web/apiserver_test.go:
```diff
@@ -862,8 +862,10 @@ func TestClusterNodesGet(t *testing.T) {
 
 	endpoint := pack.clt.Endpoint("webapi", "sites", clusterName, "nodes")
 
+	query := url.Values{"sort": []string{"name"}}
```
Not requiring things to be sorted may reduce database stress
Also, for this particular check seems to be unnecessary (we only care if exists one or none)

kimlisa replied:
it's related to this comment https://github.com/gravitational/teleport/pull/15637#discussion_r952105198, the tests that include the `sort` query params, has a test that tested for the `totalCount` which can only be returned with calls to ListResources that triggers fake pagination (sort being one of them)

marcoandredinis commented on lib/web/apiserver.go:
```diff
@@ -534,6 +534,7 @@ func (h *Handler) bindDefaultEndpoints(challengeLimiter *limiter.RateLimiter) {
 	h.GET("/scripts/:token/install-app.sh", httplib.MakeHandler(h.getAppJoinScriptHandle))
 	// web context
 	h.GET("/webapi/sites/:site/context", h.WithClusterAuth(h.getUserContext))
+	h.GET("/webapi/sites/:site/resources/check", h.WithClusterAuth(h.checkAccessToRegisteredResource))
```
:nail_care: 

`check` seems too generic but I don't have a better suggestion
Sorry, naming is hard :bow:

kimlisa replied:
i was actually hoping someone had a better name but alas :sweat_smile:

marcoandredinis reviewed (approved):
Overall looks good :+1: 
Just left some question

kimlisa committed:
Remove setting NeedTotalCount flag for web api list resources

This flag is not needed since needing total count request
will always be requested with the sort param always set which
will also return the total count. It is also being removed
because some requests coming through this layer will not require
fake pagination (which makes request slower).

kimlisa commented on lib/web/resources.go:
```diff
@@ -340,6 +378,12 @@ type listResourcesGetResponse struct {
+type checkAccessToRegisteredResourceResponse struct {
+	// HasResource is a flag to indicate if user has any access
+	// to a registered resource or not.
+	HasResource bool `json:"hasResource"`
```
i don't think this is necessary, if this endpoint returns true we can use the perms acquired by fetching user context to determine which first tab the user has full acess to

kimlisa committed:
Fix 'Incorrect conversion between integer types' from codeQL

kimlisa committed:
Merge branch 'master' into lisa/discover-resource-check
